### ISSUE TITLE
Function array arguments cannot be correctly shadowed in the compiler

### ISSUE THREAD

**Issue by Yag000** [Yag000/monkey-rs#8]
## Description

When shadowing an array function argument using he compiler I get an error. This is not present on the interpreter.

## Steps to reproduce

Run the following program:
```monkey
let func = fn (table){
    let table = table[len(table) - 1];
    table
};
puts(func([1,2,3]));
```

## Expected behavior
```monkey
3
```

## Actual behavior

```monkey
Error: RuntimeError { error: "Unsupported types for binary operation" }
```

## Additional information

This does not happen if we introduce a new variable:

```monkey
let func = fn (table){
    let new_table = table[len(table) - 1];
    new_table
};
puts(func([1,2,3]));
```

Another case is when running the iterative fibonacci example (it is how I discovered it). It probably also happens with hashes but I have not tried it yet.

**Comment by Yag000** [comment]
I think that I will release the first version  (`0.0.0`) after this patch. This is the sign of a bad implementation of the compiler and should be addressed as soon as possible. After that, the normal flow of development may begin.

**Comment by Yag000** [comment]
I've looked at this a little bit more and it turns out ( I actually expected this but I did not say it on the bug report just in case) that the issue has nothing to do with the function arguments. It is actually way more important. The issue is that the compiler cannot handle the use of the same variable both on the name and the definition. For example the following monkey code evaluates to `NULL` instead of `1`
```monkey
let a = 1;
let a = a;
a
```
Any more complex variation of this results in an error. For example. 
```monkey
let a = 1;
let a = a + 1;
```
Evaluates to an error: `Unsupported types for binary operation`, meaning that `a` is `NULL`. 
I will look into this tomorrow, for now I will push the failing tests to a new branch, in order to have something concrete to work with.
